### PR TITLE
prov/efa : implement EFA core provider's RMA read

### PR DESCRIPTION
**Pull request by wzamazon**
This patch implement EFA core provider's RMA read using rdma-core WR API.

Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

**Comment by wzamazon** [comment]
rebased to master and adjusted commit message.

File: prov/efa/src/efa_ep.c [hunk 1]
```diff
@@ -91,29 +91,37 @@ static int efa_ep_modify_qp_rst2rts(struct efa_qp *qp)
...
 }
 
-static int efa_ep_create_qp(struct efa_ep *ep,
-			    struct ibv_pd *ibv_pd,
-			    struct ibv_qp_init_attr *init_attr)
+static int efa_ep_create_qp_ex(struct efa_ep *ep,
+			       struct ibv_pd *ibv_pd,
+			       struct ibv_qp_init_attr_ex *init_attr_ex)
 {
 	struct efa_domain *domain = ep->domain;
```
Comment (zhngaj) [inline prov/efa/src/efa_ep.c]: I would separate the declaration from the assignment.

File: prov/efa/src/efa_ep.c [hunk 2]
```diff
@@ -91,29 +91,37 @@ static int efa_ep_modify_qp_rst2rts(struct efa_qp *qp)
...
 	struct efa_qp *qp;
+	struct efadv_qp_init_attr efa_attr = {};
+
```
Comment (zhngaj) [inline prov/efa/src/efa_ep.c]: NIT: This blank line can be removed.

File: prov/efa/src/efa_rma.c [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <ofi_mem.h>
+#include <ofi_iov.h>
+#include "efa.h"
+
+static
+ssize_t efa_rma_post_read(struct efa_ep *ep, const struct fi_msg_rma *msg, uint64_t flags)
+{
+	struct efa_qp *qp = ep->qp;
```
Comment (zhngaj) [inline prov/efa/src/efa_rma.c]: I would separate the declaration from the assignment.

File: prov/efa/src/efa_rma.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	struct efa_conn *conn;
+	int lkey;
+
+	assert(msg->iov_count == 1);
```
Comment (zhngaj) [inline prov/efa/src/efa_rma.c]: I might be not involved in the context. Why do we have such an assertion? Does this mean the `iov_count` in function `efa_rma_readv` needs to be 1?

File: prov/efa/src/efa_ep.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/efa_ep.c]: Good point. Fixed in new revision

File: prov/efa/src/efa_ep.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/efa_ep.c]: Thank you! removed

File: prov/efa/src/efa_rma.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/efa_rma.c]: Fixed

File: prov/efa/src/efa_rma.c [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ssize_t efa_rma_post_read(struct efa_ep *ep, const struct fi_msg_rma *msg, uint64_t flags)
+{
+	struct efa_qp *qp;
+	struct efa_conn *conn;
+	int lkey;
+
+	assert(msg->iov_count == 1);
+	assert(msg->rma_iov_count == 1);
+	assert(ep->domain->ctx->max_wr_rdma_sge >= 1);
+	/* caller must provide desc because efa claim FI_MR_LOCAL */
```
Comment (rwespetal) [inline prov/efa/src/efa_rma.c]: requires instead of claim

File: prov/efa/src/efa_ep.c [hunk 6]
```diff
@@ -91,29 +91,37 @@ static int efa_ep_modify_qp_rst2rts(struct efa_qp *qp)
...
+			       struct ibv_pd *ibv_pd,
+			       struct ibv_qp_init_attr_ex *init_attr_ex)
 {
-	struct efa_domain *domain = ep->domain;
+	struct efa_domain *domain;
 	struct efa_qp *qp;
+	struct efadv_qp_init_attr efa_attr = {};
 	int err;
 
+	domain = ep->domain;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_ep.c]: It was used in line 132
```
 domain->qp_table[ep->qp->qp_num & domain->qp_table_sz_m1] = ep->qp;
```

[the remark this answers is not in the document]
Comment (rwespetal) [reply prov/efa/src/efa_ep.c]: Ah, sorry missed that.

File: prov/efa/src/efa_rma.c (hunk 5 above)
Comment (wzamazon) [reply prov/efa/src/efa_rma.c]: Thank you! fixed

File: prov/efa/src/efa_rma.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/efa_rma.c]: In new revision, I changed assertion to check.

File: prov/efa/src/efa_rma.c [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	if (msg->rma_iov_count > ep->domain->ctx->max_wr_rdma_sge) {
+		EFA_WARN(FI_LOG_CQ, "invalid rma_iov_count!\n");
+		return -FI_EINVAL;
+	}
+
+	/* caller must provide desc because EFA require FI_MR_LOCAL */
+	assert(msg->desc);
+	assert(ofi_total_iov_len(msg->msg_iov, msg->iov_count) <= ep->domain->ctx->max_rdma_size);
```
Comment (rwespetal) [inline prov/efa/src/efa_rma.c]: Does the NIC validate size? If not, this should probably be a check as well.

File: prov/efa/src/efa_rma.c [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+#include "efa.h"
+
+static
+ssize_t efa_rma_post_read(struct efa_ep *ep, const struct fi_msg_rma *msg, uint64_t flags)
+{
+	struct efa_qp *qp;
+	struct efa_conn *conn;
+	int lkey;
+
+	if (msg->iov_count > ep->domain->ctx->max_wr_rdma_sge) {
```
Comment (rwespetal) [inline prov/efa/src/efa_rma.c]: Minor thing: these should probably be wrapped with `OFI_UNLIKELY`

Comment (wzamazon) [reply prov/efa/src/efa_rma.c]: Fixed in new revision

File: prov/efa/src/efa_rma.c (hunk 7 above)
Comment (wzamazon) [reply prov/efa/src/efa_rma.c]: Changed to check in new revision

**Pull request merged**